compiler/codegen/OMRRegisterIterator.hpp

```diff
@@ -22,36 +22,61 @@
 #ifndef OMR_REGISTER_ITERATOR_INCL
 #define OMR_REGISTER_ITERATOR_INCL
 
-/*
- * The following #define and typedef must appear before any #includes in this file
- */
-#ifndef OMR_REGISTER_ITERATOR_CONNECTOR
-#define OMR_REGISTER_ITERATOR_CONNECTOR
-namespace OMR { class RegisterIterator; }
-namespace OMR { typedef OMR::RegisterIterator RegisterIteratorConnector; }
-#endif
-
 #include "env/TRMemory.hpp"  // for TR_Memory, etc
+
+namespace TR { class Machine; }
 namespace TR { class Register; }
 
 namespace OMR
```

Comment:
Given `RegisterIterator` is no longer an extensible class, I think its implementation should be moved to the `TR` namespace, like the other non-extensible classes.

Reply:
@Leonardo2718 Initially I've implemented `RegisterIterator` in the `TR` namespace but @0xdaryl suggested to move it into the `OMR` one (please have a look at https://github.com/eclipse/omr/pull/2981#discussion_r222771376) Fortunately, I've used a separate commit for these changes, so if you think the class must be in the `OMR` namespace, I can just reset the commit.

@0xdaryl @Leonardo2718 @fjeremic Which namespace should be the home for `RegisterIterator`?

Reply:
Hm... alright.

According to our current documentation in [Extensible_Classes_in_OMR.md](https://github.com/eclipse/omr/blob/master/doc/compiler/extensible_classes/Extensible_Classes_in_OMR.md#namespaces) (emphasis mine):

>We use the TR:: namespace to serve this purpose. Outside of the layered construction of an extensible class, all compiler code uses the TR:: namespace (rooted in the code name for the compiler technology which was Testasrossa).
>
>In the case of extensible classes, we expose them to consumers by providing a terminating subclass in the TR:: namespace, which we call the Concrete class.
>
>**In the case of classes that aren't extensible, It is perfectly legitimate to create a class directly in the TR:: namespace if it's complete and can and should be used as-is**.

Has there been some discussion to change this that I've missed? I see #2084 is still open, but it hasn't had any activity in quit some time. :slightly_smiling_face:

Reply:
I thought `TR::` was reserved _for_ extensible classes and `OMR::` for non-extensible. I think @samolisov's change here is correct.

Reply:
@Leonardo2718 do you agree with @fjeremic about the namespace?

Reply:
It seems like a deeper discussion is needed to figure out what namespace we want non-extensible classes to be in, so I've opened #3306.

In the mean time, I think this PR shouldn't have to wait for that discussion to conclude and should just follow our currently documented guideline, which is to use the `TR` namespace (at least, that's how I'm interpreting the paragraph I highlighted above).

@fjeremic does this sound reasonable?

Reply:
@Leonardo2718 @fjeremic OK, I've moved `RegisterIterator` back to the `TR` namespace. The class may be moved to the `OMR` namespace as well as all the related ones when the decision is made. If you have no more objections, please approve the PR.